src/MongoDB/WriteConcern.c

```diff
@@ -329,9 +314,7 @@ static HashTable* php_phongo_write_concern_get_properties_hash(zval* object, boo
 
 	wtag     = mongoc_write_concern_get_wtag(intern->write_concern);
 	w        = mongoc_write_concern_get_w(intern->write_concern);
-	/* Note: PHP currently enforces that wimeoutMS is a 32-bit integer, so
-	 * casting will never truncate the value. This may change with PHPC-1411. */
-	wtimeout = (int32_t) mongoc_write_concern_get_wtimeout_int64(intern->write_concern);
+	wtimeout = mongoc_write_concern_get_wtimeout_int64(intern->write_concern);
```

Comment:
Sorry for not catching this earlier, but is there a concern with silent truncation affecting BSON serialization via `bsonSerialize()`, which uses this `get_properties_hash` method? If so, we may want to consider emitting an Int64 object for an out-of-range wtimeout value, on par with emitting a string in `php_phongo_write_concern_to_zval`. AFAICT, the `to_zval` function is only used for BulkWrite's debug info.

Also, `PHP_METHOD(WriteConcern, serialize)` still includes the code being removed here:

```c
/* Note: PHP currently enforces that wimeoutMS is a 32-bit integer, so
 * casting will never truncate the value. This may change with PHPC-1411. */
wtimeout = (int32_t) mongoc_write_concern_get_wtimeout_int64(intern->write_concern);
```

I did some digging and realized that most of our value objects (BSON types and the RC/RP/WC classes) use their `init_from_hash` methods for both `_set_state` _and_ `unserialize` methods; however, the `serialize` methods _do not_ utilize the `get_properties_hash` function. `get_properties_hash` tends to only be used for the `get_debug_info` and `get_properties` handlers. This means that there's some subtle code duplication between `get_properties_hash` and `serialize`. I wonder if it's worth trying to refactor this in some future ticket.

In the meantime, I think we should ensure that WriteConcern's `serialize` method remains consistent with `get_properties_hash`. I wasn't too worried about `getWtimeout()` truncating, since the alternative is a BC break, but I think we can ensure compatibility in our debugging and serialization output (both PHP _and_ `bsonSerialize()`) by conditionally emitting an `Int64` object. Thoughts?

I'm happy to punt on this for the time being and create a follow-up ticket. It's something we could certainly address in a later minor version, since this is admittedly an edge case. I believe the only users affected by this would be those that manage to create a 64-bit wtimeout on a 32-bit system and want to have that WriteConcern serialized properly in a command document by way of `bsonSerialize()`. I'm not sure about how a 32-bit PHP instance would handle encountering a 64-bit integer in a serialized string -- presumably there'd be some error/truncation well before we even call `php_phongo_writeconcern_init_from_hash` from `PHP_METHOD(WriteConcern, unserialize)`.

Reply:
That's a very good point.

I've ensured that we don't truncate in any place other than `getWtimeout`, where we have to do so to maintain backward compatibility. `getWtimeout` will now raise an `E_WARNING` when a value outside of the 32-bit range is encountered on a 32-bit system (see separate commit).

The logic in `serialize` and `get_properties_hash` have been adjusted: the former will dump a 64-bit value as string when encountering a 64-bit value, while the latter will create a `MongoDB\BSON\Int64` object and return that instead.

The logic in `unserialize` needs adjusting: the value is currently required to be a `long`, which we'll have to adapt to the changes in `serialize`. I just wanted to double-check that this is indeed feasible before going down that route.